src/scp/BallotProtocol.cpp

```diff
@@ -1334,99 +1334,96 @@ BallotProtocol::setAcceptCommit(SCPBallot const& c, SCPBallot const& h)
-            {
-                eCounter = UINT32_MAX;
-            }
-            break;
-            default:
-                abort();
-            };
-            if (eCounter >= targetCounter)
-            {
-                allCounters.insert(eCounter);
-            }
+            uint32_t c = statementBallotCounter(e.second.statement);
+            if (c >= localCounter)
+                allCounters.insert(c);
         }
 
-        // uses 0 as a way to track if a v-blocking set is at a higher counter
-        // if so, we move to that smallest counter
-        allCounters.insert(targetCounter);
-
-        // go through the counters, find the smallest not v-blocking
-        for (auto it = allCounters.begin(); it != allCounters.end(); it++)
+        // If we got to here, implicitly there _was_ a v-blocking subset
+        // with counters above the local counter; we just need to find a
+        // minimal n at which that's no longer true. So check them in
+        // order, starting from the smallest.
+        for (uint32_t n : allCounters)
         {
-            uint32 n = *it;
-
-            bool vBlocking = LocalNode::isVBlocking(
-                getLocalNode()->getQuorumSet(), mLatestEnvelopes,
-                [&](SCPStatement const& st) {
-                    bool res;
-                    auto const& pl = st.pledges;
-                    if (pl.type() == SCP_ST_PREPARE)
-                    {
-                        auto const& p = pl.prepare();
-                        res = n < p.ballot.counter;
-                    }
-                    else
-                    {
-                        if (pl.type() == SCP_ST_CONFIRM)
-                        {
-                            res = n < pl.confirm().ballot.counter;
-                        }
-                        else
-                        {
-                            res = n != UINT32_MAX;
-                        }
-                    }
-                    return res;
-                });
-
-            if (n == targetCounter)
+            if (!hasVBlockingSubsetStrictlyAheadOf(localNode, mLatestEnvelopes,
+                                                   n))
             {
-                // if current counter is not behind, don't do anything
-                if (!vBlocking)
-                {
-                    break;
-                }
-            }
-            else
-            {
-                if (!vBlocking)
-                {
-                    // move to n
-                    return abandonBallot(n);
-                }
+                // Move to n.
+                return abandonBallot(n);
             }
         }
-    }
 
+        // We should only have got here if allCounters was empty; otherwise
+        // there should at worst have been some _maximal_ counter in the set to
+        // advance to, that by definition no v-blocking subset was "strictly
+        // ahead of".
+        assert(allCounters.empty());
```

Comment:
I don't think this is actually reachable. Suppose that there exists a v-blocking set strictly ahead of `localCounter`. Then there exists at least one element in `mLatestEnvelopes` with counter `c > localCounter`, because the empty set is not a v-blocking set. Therefore `allCounters` contains a counter `c > localCounter`; specifically `allCounters` is not empty. Now fix the highest counter `C` in `allCounters`. If there does not exist a v-blocking set strictly ahead of `C`, then this line cannot be reached and we are done. Suppose instead that there does exist a v-blocking set strictly ahead of `C`. Then, as before, there exists at least one element in `mLatestEnvelopes` with counter `c > C`, because the empty set is not a v-blocking set. But this implies that `allCounters` includes `c > C`, which contradicts that `C` is the highest counter in `allCounters`.

Reply:
I agree, this comment and assert are confusing, I am removing them. I could have put an `assert(false)` here but this doesn't seem to meet the bar for crashing consensus